core/src/protosext/mod.rs

```diff
@@ -223,6 +223,25 @@ impl HistoryEventExt for HistoryEvent {
     }
 }
 
+pub(crate) trait FailureExt {
+    /// Extracts an ApplicationFailureInfo from a Failure instance if it exists
+    fn maybe_application_failure(&self) -> Option<&ApplicationFailureInfo>;
+}
+
+impl FailureExt for Failure {
+    fn maybe_application_failure(&self) -> Option<&ApplicationFailureInfo> {
+        if let Failure {
+            failure_info: Some(FailureInfo::ApplicationFailureInfo(f)),
+            ..
+        } = self
+        {
+            Some(f)
+        } else {
+            None
+        }
+    }
+}
+
```

Comment:
This can go as a direct implementation in the protos crate

Reply:
Missed that, thanks